controller: wfa_ca: add validation checks for input

CoralMalachi opened the pull request:
The MCUT (MultiAP Controller Under Test) gets a string command from
WTS UCC (Wifi Test Suite Unified CAPI Console).

The `tlv_value` fields' input should be inside curly brackets
and in hexadecimal format.

The following changes consist of two validation checks:
1.Check for balanced parentheses input (only curly brackets).
2.Check if value string conforms to hex notation.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller:wfa_ca: add validation checks for input

The MCUT (MultiAP Controller Under Test) gets a string command from
WTS UCC (Wifi Test Suite Unified CAPI Console).

The `tlv_value` fields' input should be inside curly brackets
and in hexadecimal format.

The following changes consist of two validation checks:
1.Check for balanced parentheses input (only curly brackets).
2.Check if value string conforms to hex notation.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -67,6 +67,12 @@ const std::string wfa_ca::wfa_ca_status_to_string(eWfaCaStatus status)
     return std::string();
 }
 
+static bool is_hex_notation(std::string const &s)
```
I would remove the "is" prefix.
Also, I think te convention in our code is to put the const before the type (nitpick)

tomereli replied:
Agree on the const, disagree about the is

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                    prev_curly_bracket = true;
+                    prev_space         = false;
+                    count_open_brackets++;
+                } else {
```
You could avoid nesting by using "continue":
```
if (c == '{') {
    // code
    continue;
}

// rest of the code
```

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+
+                    if (count_open_brackets == 0) {
+                        err_string =
+                            "The following string is not valid (curly brackets test failed) : '" +
```
I would change it to something like "Unbalanced curly brackets in value of param '<param_name>'"
No need to mention that the string is invalid because the return code is "INVALID".

CoralMalachi replied:
Ok, I'll change this .

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                        count_open_brackets--;
+                        if (count_open_brackets < 0) {
+
+                            err_string = "The following string is not valid (curly brackets test "
```
same here

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                        prev_space = true;
+
+                    } else {
+                        if (prev_space == true || prev_curly_bracket == true) {
```
Here you could also avoid nesting:
```
if (!condition) {
    // continue;
}
// rest of the code
```

morantr replied:
IMO the comparison to true/false of booleans is not necessary.
if (VAR == true) -> if (VAR)
if (VAR == false) -> if (!VAR)

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                            auto token =
+                                command_str.substr(i, command_str.find_first_of(" }", i) - i);
+                            token.pop_back();
+                            //Check if a givven string conforms to hex notation.
```
givven -> given

morantr reviewed (changes requested):
Regarding the brackets balance check, I see what you did there, but I think it is over complicated.
I would instead split the string by spaces and get a vector for example:
"{0x00904C2A11C2", "0x14", "{0x51", "{0x00", "0x00}}"
Check that the first element starts with '{' and the last one end s with '}'
Then, use std::count() to count how many '{' and '}' to create a balance of the brackets,
if the balance if negative return error.
Then check the hex notation of the value.
I think it will be more easy to read.

Second, I think this should be in a static file function.

Also, I know that I thought that it should be checked in the parse function, but yesterday I pushed a commit that handles specifically dev_send_1905 command, and I think it will be more fit to check it the function get_send_1905_1_tlv_hex_list() of the followup commit:
https://github.com/prplfoundation/prplMesh/pull/180/commits/f8bd822b5189616984d97d1e3c6a5746e2d9b9c9

So meanwhile fix the PR, and split it brackets balance check to a different function, and when my commit will be merged, use it in the correct places.

CoralMalachi commented:
> Regarding the brackets balance check, I see what you did there, but I think it is over complicated.
> I would instead split the string by spaces and get a vector for example:
> "{0x00904C2A11C2", "0x14", "{0x51", "{0x00", "0x00}}"
> Check that the first element starts with '{' and the last one end s with '}'
> Then, use std::count() to count how many '{' and '}' to create a balance of the brackets,
> if the balance if negative return error.
> Then check the hex notation of the value.
> I think it will be more easy to read.
> 
> Second, I think this should be in a static file function.
> 
> Also, I know that I thought that it should be checked in the parse function, but yesterday I pushed a commit that handles specifically dev_send_1905 command, and I think it will be more fit to check it the function get_send_1905_1_tlv_hex_list() of the followup commit:
> [f8bd822](https://github.com/prplfoundation/prplMesh/commit/f8bd822b5189616984d97d1e3c6a5746e2d9b9c9)
> 
> So meanwhile fix the PR, and split it brackets balance check to a different function, and when my commit will be merged, use it in the correct places.

Do you mean to do the brackets balance check first and just after that do the hex notation check?

arnout commented:
When looking at #180, I actually wonder if it really makes sense to have a separate check. The conversion from hex to uint8_t implicitly verifies that it really is a hex number.

As for the curly braces: we don't really care about them. It is not a requirement that we fail if the curly braces matching is done incorrectly on input. For us, it is sufficient to just remove them, like is done in #180.

@tomereli @morantr  please comment.

morantr commented:
> When looking at #180, I actually wonder if it really makes sense to have a separate check. The conversion from hex to uint8_t implicitly verifies that it really is a hex number.
> 
> As for the curly braces: we don't really care about them. It is not a requirement that we fail if the curly braces matching is done incorrectly on input. For us, it is sufficient to just remove them, like is done in #180.
> 
> @tomereli @morantr please comment.

The hex notation check is definitely necessary since in #180 I use std::stoi() that if it will fail, it will throw an exception, so it is a good idea to do add hex notation check before converting it to an integer.
Regarding, the curly bracket balance check, yes it is a "nice to have check", and the question is if we want to validate that input.

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -147,6 +153,7 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
                           std::unordered_map<std::string, std::string> &params,
                           std::string &err_string)
 {
+
```
Extra line break

tomereli commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -157,7 +164,9 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
         std::transform(param_name.begin(), param_name.end(), param_name.begin(), ::tolower);
+
         if (params.find(param_name) != params.end()) {
+
```
Here as well

simonbaren commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                    err_string = "Unbalanced curly brackets in value of param " + param_name + "'";
+                    return false;
+                }
+                if (c == '}') {
```
else if

simonbaren commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -166,6 +175,60 @@ bool wfa_ca::parse_params(const std::vector<std::string> &command_tokens,
+                    prev_curly_bracket = false;
+                    continue;
+                }
+                if (c == ' ') {
```
else if

morantr reviewed (changes requested):
I'm not approving it now. There are still changes need to be done.
Also, the validation check could be in a more proper place after #180 will be merged.
@tomereli don't rush to merge it.

Commit:
fixup! controller:wfa_ca: add validation checks

Split the brackets balance check to a different function to make the
code more easy to read.

arnout commented:
As discussed, these checks got included in #180.
